include/mfmg/cuda/cuda_matrix_free_operator.cuh

```diff
@@ -0,0 +1,75 @@
+/*************************************************************************
+ * Copyright (c) 2017-2019 by the mfmg authors                           *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#ifndef MFMG_CUDA_MATRIX_FREE_OPERATOR_CUH
+#define MFMG_CUDA_MATRIX_FREE_OPERATOR_CUH
+
+#ifdef MFMG_WITH_CUDA
+
+#include <mfmg/common/operator.hpp>
+#include <mfmg/cuda/cuda_handle.cuh>
+#include <mfmg/cuda/cuda_matrix_free_mesh_evaluator.cuh>
+
+namespace mfmg
+{
+template <int dim, typename VectorType>
+class CudaMatrixFreeOperator final : public Operator<VectorType>
+{
+public:
+  using size_type = typename VectorType::size_type;
+  using value_type = typename VectorType::value_type;
+  using vector_type = VectorType;
+
+  CudaMatrixFreeOperator(std::shared_ptr<CudaMatrixFreeMeshEvaluator<dim>>
+                             matrix_free_mesh_evaluator);
+
+  void vmult(vector_type &dst, vector_type const &src) const;
+
+  // Needed because the extract_row works only on the host
+  void apply(dealii::LinearAlgebra::distributed::Vector<
+                 value_type, dealii::MemorySpace::Host> const &x,
+             dealii::LinearAlgebra::distributed::Vector<
+                 value_type, dealii::MemorySpace::Host> &y,
+             OperatorMode mode = OperatorMode::NO_TRANS) const;
+
+  virtual void apply(vector_type const &x, vector_type &y,
+                     OperatorMode mode = OperatorMode::NO_TRANS) const override;
```

Comment:
I am getting confused but the coexistence of `vmult()` and `apply()` overload that take `vector_type` and the `apply()` overload that takes vectors on the host.  Was the latter introduced so you can reuse `matrix_transpose_matrix_multiply()`?

Reply:
Yes exactly. I'll add a comment.